build.sh

```diff
@@ -1,66 +1,102 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
 
+# Build dependencies for scripts in bin/
+
+set -x
+
+# Local directories for installing Python and R packages.
+R_LIB_DIR=`pwd`/work/rlibs
+PIP_TARGET_DIR=`pwd`/work/pylibs
+
+# Check requirements.
 which git > /dev/null 2>&1
-if [ $? -ne 0 ]; then
+if [ $? != 0 ]; then
     echo "git must be installed" >&2
     exit 1
 fi
 
 which python2.7 > /dev/null 2>&1
-if [ $? -ne 0 ]; then
+if [ $? != 0 ]; then
     echo "python2.7 must be installed" >&2
     exit 1
 fi
 
 python2.7 -c "import setuptools" > /dev/null 2>&1
-if [ $? -ne 0 ]; then
+if [ $? != 0 ]; then
     echo "setuptools for python2.7 must be installed" >&2
     exit 1
 fi
 
-set -e
-
-mkdir -p work
-R_INST_DIR=`pwd`/work/R-inst
-PIP_TARGET_DIR=`pwd`/work/pylibs
-
-# Download a recent version of R (must be 3.3.1 or later).
-mkdir -p ${R_INST_DIR}
-cd work
-wget https://cran.r-project.org/src/base/R-3/R-3.3.2.tar.gz
-tar -xzf R-3.3.2.tar.gz
-rm R-3.3.2.tar.gz
-cd R-3.3.2
+which R > /dev/null 2>&1
+if [ $? != 0 ]; then
+    echo "R version 3.3.1 or later must be installed"
+    exit 1
+fi
 
-# Build R.
-./configure --prefix=${R_INST_DIR} --enable-R-shlib
-make
-make install
-cd ..
-rm -Rf R-3.3.2/
-cd ..
+version() {
+    echo "$@" | awk -F. '{ printf("%d%03d%03d%03d\n", $1,$2,$3,$4); }';
+}
 
-# Correct PATH and LD_LIBRARY_PATH variables.
-export PATH="${R_INST_DIR}/bin:${PATH}"
-export LD_LIBRARY_PATH="${R_INST_DIR}/lib/R/lib:${LD_LIBRARY_PATH}"
+if [ $(version `R --version | awk 'NR==1 {print $3}'`) <= $(version "3.3.1") ]; then
+    echo "R version 3.3.1 or later must be installed"
+    exit 1
+fi
 
 # Install R changepoint package.
-echo "install.packages('devtools', lib='${R_INST_DIR}/lib/R/library', repos='http://cran.us.r-project.org')" | R_LIBS_USER=${R_INST_DIR}/lib/R/library/ ${R_INST_DIR}/bin/R --no-save
-echo "install.packages('MultinomialCI', lib='${R_INST_DIR}/lib/R/library', repos='http://cran.us.r-project.org')" | R_LIBS_USER=${R_INST_DIR}/lib/R/library/ ${R_INST_DIR}/bin/R --no-save
-echo "devtools::install_git('git://github.com/rkillick/changepoint', branch = 'master')" | R_LIBS_USER=${R_INST_DIR}/lib/R/library/ ${R_INST_DIR}/bin/R --no-save
+mkdir -p ${R_LIB_DIR}
+echo "install.packages('devtools', lib='${R_LIB_DIR}', repos='http://cran.us.r-project.org')" | R_LIBS_USER=${R_LIB_DIR} R --no-save
+echo "install.packages('MultinomialCI', lib='${R_LIB_DIR}', repos='http://cran.us.r-project.org')" | R_LIBS_USER=${R_LIB_DIR} R --no-save
+echo "devtools::install_git('git://github.com/rkillick/changepoint', branch = 'master')" | R_LIBS_USER=${R_LIB_DIR} R --no-save
 
 # Install rpy2 Python package.
 DEB8_HACK=0
 if [ -f /etc/debian_version ]; then
     DEB_MAJOR_V=`cat /etc/debian_version | cut -d. -f1`
-    if [ "${DEB_MAJOR_V}" = "8" ]; then
+    if [ "${DEB_MAJOR_V}" == "8" ] || [ "${DEB_MAJOR_V}" == "9" ]; then
```

Comment:
I'm not sure this syntax is correct. Dos it not work with `if [ "${DEB_MAJOR_V}" == "8" || "${DEB_MAJOR_V}" == "9" ]; then`?

Reply:
Single equals should work since both sides are quoted string literals.

But more importantly, the script expects `/etc/debian_version` to contain a debian version number, but on ubuntu it seems to contain the codename for the release, so the comparison is broken anyway.

On bencher8:
```
$ cat /etc/debian_version 
9.4
```

Reply:
@snim2 does your ubuntu machine have a `/etc/os-release`?

Reply:
The syntax @ltratt suggests errors on my machine, and I don't see any of the errors @fsfod sees, but I do now see this: `./build.sh: 77: ./build.sh: cannot open =: No such file`

My Ubuntu machine has this in `/etc/os-release`:

```sh
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="18.04.1 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.1 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

Reply:
@snim2 What errors do you get?

Reply:
Only the one above.

Reply:
Here's a debian `/etc/os_version`:
```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

So instead of looking at `/etc/debian_version` you can look at the `ID` and `VERSION_ID` fields and base the check on those:

```
$ awk -F= '/^ID=/ {print "\"" $2 "\""}' /etc/os-release 
"debian"
$ awk -F= '/^VERSION_ID=/ {print $2}' /etc/os-release 
"9"
```

Reply:
Would `uname -v` be easier to parse?

Reply:
Ugh, I see on Bencher 8 I get similar errors to the ones @fsfod sees `[: 9: unexpected operator` etc. OK, let me see if your idea works.